### PR TITLE
Address fragment pasting in nested documents in a more general manner.

### PR DESCRIPTION
**Pull request by bebbi**
#### Is this adding or improving a _feature_ or fixing a _bug_?

This addresses #2575 .

#### What's the new behavior?

Same behaviour for examples and tests.

#### How does this change work?

The proposed change address fragment pasting in nested documents in a more general manner.

Here is an explanation to visualize this change:
Think of the leaves of the document as the document's surface, and think of the paste fragment as a wedge. The objective is to drive the wedge as deeply into the surface as reasonable to maximize smoothness of the resulting document surface.

The constraints from the previous implementation still apply: we can only go so deep that the ancestor node type at each ancestor level of the wedge matches the document's ancestor node type at some corresponding depth level.

To achieve this, we make our "wedge depth" calculation relative to document surface instead of document root.

Unlike the current document root based calculation, this ensures we won't be trying to match our wedge against nodes too deep below the surface level when we paste a shallow wedge into a deep part of the doc.

The new approach solves the issue's fiddle; it generally gives copying from shallow trees into deep trees a reasonable default behaviour that I think should be more accurate than the previous.

#### An example:

In our use-cases, we have deeply nested trees that have `container` type nodes which in turn contain blocks of leaves, and more `container`s, similar to what's in the fiddle.
But the example will looks similar if we have `table`s nested inside `table`s or other type of repeating structures.

Now if our document selection has the following ancestor structure:

    container
      blockWithTextLeaf
      container
        blockWithTextLeaf
          <cursor>

and we're trying to copy-paste a word from a paste fragment that looks like:

    container
      blockWithTextLeaf
        <word to paste>

Then the new approach lets us match the fragment container with the leaf-most container as opposed to the root-most container which wouldn't be helpful.

See https://github.com/ianstormtaylor/slate/issues/2575#issuecomment-462027027

#### Have you checked that...?

* [x] The new code matches the existing patterns and styles.
* [x] The tests pass with `yarn test`.
* [x] The linter passes with `yarn lint`. (Fix errors with `yarn prettier`.)
* [x] The relevant examples still work. (Run examples with `yarn watch`.)

#### Does this fix any issues or need any specific reviewers?

Fixes: #2575
Reviewers: @ianstormtaylor

### CODE REVIEW

**Comment by jtadmor** [comment]
We've had to implement something similar like this for lists / tables, I can try to see if this code would prevent the need for custom code to handle those use cases.

**Comment by steobrien** [comment]
I haven't got time to closely examine and think think through the implications here, but from a quick read this looks great.

However, it would be great to see some new fixtures added to demonstrate – and lock in – the improvement. In my experience for problems like these, the fixtures are more long-term valuable than the code itself.

Do you think you could isolate some examples, ideally based on the real-world problems you encountered?

**Comment by bebbi** [comment]
@steobrien yes I was thinking of getting some feedback and then add a test based on the solved fiddle, agree that this is important.

File: packages/slate/src/commands/at-range.js [hunk 1]
```diff
@@ -807,36 +807,86 @@ Commands.insertFragmentAtRange = (editor, range, fragment) => {
...
+ * @param {string} documentKey
+ * @param {Document} fragment
+ * @return {Node}
+ */
 
-  if (hasSingleNode(fragment)) {
-    let fragmentInner = firstNode(fragment)
+const findInsertionNode = (fragment, document, documentKey) => {
+  // Find the deepest block in a doc with no siblings.
+  const deepestSingleBlock = doc => {
```
Comment (jtadmor) [inline packages/slate/src/commands/at-range.js]: Note that this will return `doc` even if doc has multiple child nodes. Is that intended?

File: packages/slate/src/commands/at-range.js [hunk 2]
```diff
@@ -807,36 +807,86 @@ Commands.insertFragmentAtRange = (editor, range, fragment) => {
...
+    )
+  }
 
-      if (fragmentInner.type === documentInner.type) {
-        node = fragmentInner
-      } else {
-        break
+  // Given two reverse lists of ancestors, check if all fragment ancestor types
+  // match the doc ancestors at some position.
+  const findMatchIndex = (documentAncestors, fragmentAncestors) => {
```
Comment (jtadmor) [inline packages/slate/src/commands/at-range.js]: Since this returns a block + children, maybe call it `findMatchAncestor` / `findMatchFragment` or some such?

File: packages/slate/src/commands/at-range.js [hunk 3]
```diff
@@ -807,36 +807,86 @@ Commands.insertFragmentAtRange = (editor, range, fragment) => {
...
+    for (let fragIdx = 0; fragIdx < fragmentAncestors.size; fragIdx++) {
+      // The docIdx loop relaxes our check in that we can still match if there
+      // are node type differences leaf-side.
+      // This is important for example if our fragment inserts multiple siblings
+      // or inserts another type while the tree structure remains the same.
+      for (
+        let docIdx = 0;
+        docIdx <= documentAncestors.size - fragmentAncestors.size;
```
Comment (jtadmor) [inline packages/slate/src/commands/at-range.js]: What about the case of `fragmentAncestors.size` > `docmentAncestors.size`? If we are never getting a match in this case (which appears to be the case), we can at least return quickly.

**Comment by jtadmor** [comment]
Made a couple comments about potential edge cases that may need to be addressed via code and / or tests. Would generally agree (per your issue) that it would be nice if a lot of this logic could be extracted into internal utils or even some exported utils that plugin-developers could rely on.

**Comment by bebbi** [comment]
thanks for review @jtadmor !
Actually my thinking was that this PR can serve as a pretty generic implementation until someone comes up with a breaking example. It certainly covers the existing and new tests more generally than before (and at all).

By its nature and like in the previous code, it can't nicely solve some cases of pasting from deep to shallow because if the fragment is deeper than the doc, we may lack some user-land information on deciding about how planar the resulting document surface should be.
I think I have an approach to this, as well as for the general case that has no repeating ancestor types; we could leave this issue open beyond accepting the PR to discuss that second part or create a new one.

File: packages/slate/src/commands/at-range.js (hunk 3 above)
Comment (jtadmor) [reply packages/slate/src/commands/at-range.js]: Agreed, just wanted to make sure that was the functionality you had in mind.

Also, since the loop will never find anything, we can probably do a check at the top, before upper `for` loop, so that we can return fast if `fragmentAncestors.size > documentAncestors.size`.

**Comment by jtadmor** [comment]
I like the idea of getting this in sooner if tests look good and then working on more complex solutions / utils in a later PR.

File: packages/slate/src/commands/at-range.js (hunk 3 above)
Comment (bebbi) [reply packages/slate/src/commands/at-range.js]: ah, got you

**Comment by ianstormtaylor** [comment]
Hey @bebbi, thanks this sounds awesome!

I agree with @steobrien that us having good test fixtures is the way that this kind of behavior is going to be solved long term. If I were to merge this, I'd like to be able to easily see some of the common use cases for this kind of pasting in our test fixtures, otherwise it's hard for me to tell what is changed.

Off the top of my head, for a list of "structures" that we should try to support out of the box with reasonable behavior if possible:

- Individual words/phrases
- Flat paragraphs
- Nested quotes
- Nested lists
- Nested tables
- Nested "complex" blocks

We probably already have fixtures for some of these. But I think to merge this I'd like to see someone write up a quick summary of what we have now, what we'd need to be at a comfortable level, and then have the PR that we make include those extra fixtures.

**Comment by bebbi** [comment]
Rebasing this branch onto newest master.
This allows us to add in more pasting tests in this branch that would throw errors on previous base,
and to close #2604 and #2605.

Still todo: review the test cases above, do we have all these cases covered.

**Comment by jtadmor** [comment]
Hi @ianstormtaylor @bebbi @steobrien,

I re-arranged the existing `insert-fragment` into multiple folders. I renamed some of the existing tests, removed duplicate ones, and added a new one. I think this should better indicate the intent of the various tests and give us a decent picture of the current behavior.

I created `tables` and `lists` folders although strictly speaking those are just examples of `nested-block` pastes. They do give some clearer picture of how some existing behavior may not be ideal. I've added "TODO" comments to indicate where I personally think it is not.

One thing I would note is that Slate may not be able to handle all scenarios correctly. For example, unwrapping a `quote` block and merging the text into a `paragraph` may be reasonable, but doing the same behavior into a table would likely not be.

I believe that suggests that while this PR is certainly an improvement, the final state will almost certainly require that Slate either provide a boolean param for `insertFragment`, provide additional commands, or at least provide some utils for decomposing a fragment and separately inserting / merging leaf nodes,

**Comment by steobrien** [comment]
Nice work! The diff is hard to read here, maybe this needs a rebase/merge?

**Comment by jtadmor** [comment]
@bebbi @steobrien @ianstormtaylor Force pushed a rebased version of this, diff should be cleaner now

**Comment by bebbi** [comment]
Thanks, this looks like a solid improvement to me.

**Comment by steobrien** [comment]
This is looking good – great to see some new test cases! Are you planning to solve the TODOs before merging or is your intention for them to live in the codebase?

**Comment by jtadmor** [comment]
I wouldn't necessarily feel right "solving" them as part of this PR. That is how functionality works today for certain basic pastes, but I would want some more input if others similarly feel strongly that some of it is wrong.

Each one can then be solved one at a time, if the community is cool with it.

**Comment by ianstormtaylor** [comment]
Hey @bebbi and @jtadmor, thank you for this!

I've been hesitating to merge it, because of the huge changes coming in , but I don't want to lose this work. Especially appreciative of the new tests! So I'm going to merge this now, but I can't guarantee the behavior stays exactly as indicated in this PR now for the next few iterations of Slate. We might need to revisit this at some point, and really come to a good decision about how fragment insertion should work by looking at other editors. I might have to break fragment insertion just to get over the barrier of shipping some of these bigger changes. I hope that makes sense. I really appreciate the work you two have put in here.
